batcher/sweep_batcher.go

```diff
@@ -0,0 +1,589 @@
+	chainNotifier lndclient.ChainNotifierClient
+
+	// signerClient is the signer client that is used by batches.
+	signerClient lndclient.SignerClient
+
+	// musig2ServerKit includes all the required functionality to collect
+	// and verify signatures by the swap server in order to cooperatively
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// verifySchnorrSig is a function that can be used to verify a schnorr
+	// signature.
+	VerifySchnorrSig VerifySchnorrSig
+
+	// chainParams are the chain parameters of the chain that is used by
+	// batches.
+	chainParams *chaincfg.Params
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2ServerSigner MuSig2SignSweep,
+	verifySchnorrSig VerifySchnorrSig, chainparams *chaincfg.Params,
+	store BatcherStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[lntypes.Hash]*Batch),
+		SweepReqs:        make(chan SweepRequest),
+		errChan:          make(chan error),
+		completeChan:     make(chan lntypes.Hash),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		VerifySchnorrSig: verifySchnorrSig,
+		chainParams:      chainparams,
+		store:            store,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (br *Batcher) Run(ctx context.Context) error {
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then resume the execution of these batches.
+	batches, err := br.FetchUnconfirmedBatches(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		err := br.spinUpBatchFromDB(ctx, *batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweepReq := <-br.SweepReqs:
+			sweep, err := br.FetchSweep(ctx, sweepReq)
+			if err != nil {
+				return err
+			}
+
+			err = br.handleSweep(ctx, sweep, sweepReq.Notifier)
+			if err != nil {
+				return err
+			}
+
+		case id := <-br.completeChan:
+			delete(br.batches, id)
+
+		case err := <-br.errChan:
+			return err
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// handleSweep handles a sweep request by either placing it in an existing
+// batch, or by spinning up a new batch for it.
+func (br *Batcher) handleSweep(ctx context.Context, sweep *Sweep,
+	notifier SpendNotifier) error {
+
+	// When batches purge sweeps they will be returned here in order to be
+	// handled again. In that case the batch will not define a notifier as
+	// it does not care about the future outcome of the sweep. The swap has
+	// already registered a notifier with the batcher.
+	if notifier != (SpendNotifier{}) {
+		go br.monitorSpendAndNotify(ctx, sweep, notifier)
```

Comment:
We wanted to remove all the spend-monitoring logic from the loop-out codebase and have them notified on the sweep's success via a channel

Reply:
But once the batch is confirmed we could just notify through each sweep in that batch right?

Reply:
> notify through each sweep in that batch right

You mean once batch is spent, we detect all the sweeps that actually appear in the tx and then notify the swaps?
-- instead of having a spend monitor for each?